### ISSUE TITLE
Preparation of sparse compressed incidence matrix

### ISSUE THREAD

**Issue by mfrancesconi81** [churchill-lab/alntools#2]
Hi, 

I'm trying to perform ASE using EMASE. I am trying to follow the documentation of alntools to produce the sparse alignment incident matrix, but I am not quite sure on the details of the diploid BAM alignment to input in alntools.  Does it has to be an alignment to diploid transcriptome? or genome? Which aligner should I use? How should I prepare the reference diploid genome/transcriptome? Should I only output the best alignment? 

Many thanks, 

Mirko

**Comment by kbchoi-jax** [comment]
Could you explain your experimental setup? See if the following links make sense.

http://churchill-lab.github.io/g2gtools/
http://churchill-lab.github.io/emase-zero/

**Comment by mfrancesconi81** [comment]
Hi KB,

I have RNA seq of parents and F1 hybrids of highly inbred lines. I also
have DNA sequencing of the parents from which I did variant calling and
constructed personal genome sequences (using GATK variant calling pipeline,
and FastaAlternateReferenceMaker).

So if I understand correctly from the links you send me, once I got the
personal genomes, what I need is a BAM file of reads mapped to dual
transcriptome. Do you have any recommendations for that? Should I just
concatenate the two transcriptomes and run bwa for example?

Thank you very much,

Mirko

**Comment by mfrancesconi81** [comment]
Hi KB,

Also so far I just concatenated the two transcriptomes but how does the
algorithm distinguish between mapping to different chromosomes from mapping
to different haplotypes? Should I specify which are the different
haplotypes at some point?

Many thanks again,

Mirko

**Comment by narayananr** [comment]
The maternal and paternal transcriptomes should have unique IDs. For example, if we use ensembl annotation, we label maternal allele as ENSG001_M and paternal allele as ENSG001_P. That way we know the chromosomal location from ensembl ID and haplotypic origin from M/P tag. g2gtools can create these diploid transcriptome with user specified tags.

**Comment by kbchoi-jax** [comment]
EMASE has the following utility command. For `-s`, you give `L,R` or `M,P` or whatever suffix.

```
create-hybrid:
Usage:
    create-hybrid -F <fasta_files> [ -s <hap_list> -o <out_file> --create-bowtie-index ]

Input:
    -F <fasta_files> : List of fasta files (comma delimited)
    -s <hap_list>    : Names of haplotypes to be used instead (comma delimited, in the order of genomes)
    -o <out_file>    : Output file name (default: './emase.pooled.targets.fa')

Parameters:
    --help, -h            : shows this help message
    --create-bowtie-index : builds bowtie1 index```

**Comment by mfrancesconi81** [comment]
Hi Narayanan,

Which command of g2gtools should I use for this purpose?

Thanks,

M.

**Comment by mfrancesconi81** [comment]
Hi KB,

Is this utility also available for emase-zero?  I don't see it.

Or should I also install emase? I thought the emase-zero was equivalent but
faster because written in C.

Thanks,

Mirko

**Comment by kbchoi-jax** [comment]
You have to install `g2gtools` and `emase`. Use `g2gtools` to create founder-specific transcriptomes, and then `create-hybrid` to combine them. I am sorry my tools are currently a bit fragmented as some of the core tools are still in their development stage. I will soon reorganize it in a pipeline.

http://churchill-lab.github.io/g2gtools/
https://emase.readthedocs.io/en/latest/installation.html

**Comment by mfrancesconi81** [comment]
Ok thanks,

Your strategy seems to be statistically very sound but I'm finding a bit
difficult to understand how to perform all the steps. But I think with your
help I'll find the correct way.

Let's go through the whole thing:

You give an example of pipeline in the emase-zero documentation
http://churchill-lab.github.io/emase-zero/

1st step

bowtie -q -a --best --strata --sam -v 3 -m 100 bowtie.transcriptome
my_sample.fastq | samtools view -bS - > my_sample.bam

here bowtie.transcriptome is the result of g2gtools to create
founder-specific transcriptomes, and then emase create-hybrid to combine
them, correct?

But  the documentation of create-hybrid in emase is outdated and you have
different input https://emase.readthedocs.io/en/latest/usage.html  It seems
that you start from genomes and GTF files instead of starting with
transcriptomes.

Anyway I figured it out now.

As for the alignment does it have to bowtie? I have quickly tried with a
.bam produced with bwa mem and I got stuck at the next step (see below).


2nd step

alntools bam2ec -t transcripts.lengths.tsv -c 8 my_sample.bam my_sample.bin

here transcripts.lengths.tsv is this the output of create-hybrid (there is
a emase.pooled.transcriptome.info file produced by this command: is that
it?)

I tried to run this using a .bam produced by bwa mem and split using
alntools split


alntools bam2ec -t ${EMASE_DIR}'emase.pooled.targets.info' -c 14
${EMASE_DIR}'bwa_mem.bam' ${EMASE_DIR}'bwa_mem.bin'

some transcripts gave a warning then I got an error then the program was
stuck for ~1.5h. I figured out that it should be so slow so I terminated
it.

[alntools] [08/27/2018 05:16:13 PM] Unexpected target found in BAM file:
FBtr0302391

[alntools] [08/27/2018 05:16:13 PM] Unexpected target found in BAM file:
FBtr0089257

[alntools] [08/27/2018 05:16:13 PM] Unexpected target found in BAM file:
FBtr0301507

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0075322

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0345088

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0344894

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0333168

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0077586

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0085221

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0340116

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0334638

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0080833

[alntools] [08/27/2018 05:16:14 PM] Unexpected target found in BAM file:
FBtr0085912

[alntools] [08/27/2018 05:16:20 PM] Error: sequence item 0: expected
string, NoneType found



Do you have any clue why this happen?


$ emase-zero -m 2 -b my_sample.bin -o emase-zero.quantified -l
transcript.lengths.tsv -g gene2transcripts.tsv

Here gene2transcripts.tsv is the file produced by prepare-emase on the
reference genome, correct?

Thank you very much,

Mirko

**Comment by kbchoi-jax** [comment]
`bowtie.transcriptome` is the basename of `bowtie` index files, which you create with pooled transcriptome fasta file using `g2gtools`. See the documentation of `bowtie` if you need more detail about running `bowtie`. Are you using the most recent `emase`? Here's how to run `create-hybrid`.

```
create-hybrid:
Usage:
    create-hybrid -F <fasta_files> [ -s <hap_list> -o <out_file> --create-bowtie-index ]

Input:
    -F <fasta_files>    : List of fasta files (comma delimited)
    -s <hap_list>       : Names of haplotypes to be used instead (comma delimited, in the order of genomes)
    -A <poly_A length>  : The length of poly-A tail to append at the end of transcript sequences (default: 0)
    -o <out_file>       : Output file name (default: './emase.pooled.targets.fa')
    --delim <delimiter> : Delimiter string for new target sequence IDs (default: '_')

Parameters:
    --help, -h            : shows this help message
    --create-bowtie-index : builds bowtie1 index
```

Please show me first 10 entries of your bam file.

**Comment by mfrancesconi81** [comment]
KB,

Yes, I got it. I have now run the create-hybrid on the personal
transcriptomes created with g2gtools, including the option
create-bowtie-index. this comand is succesful in creating the
transcriptome.fa the .info  and it creates the 4 bowtie indexes (.ebwt) but
not the .rev indexes.   I then run bowtie but it complains the .rev indexes
are missing.

M.

**Comment by kbchoi-jax** [comment]
That's an error from `bowtie`. Try `bowtie-build` directly using the transcriptome.fa file. [Here](http://bowtie-bio.sourceforge.net/manual.shtml#the-bowtie-build-indexer)'s the manual.

**Comment by mfrancesconi81** [comment]
Now I get a segmentation fault error...but I coud get bowtie2 and bwa-mem
to work (both indexes and alignment). Do I have to necessarily use bowtie1?
can't I use bowtie2 or bwa mem?

thanks,


Mirko

**Comment by kbchoi-jax** [comment]
Most EM algorithms, including `emase`, do not consider how many mismatches each alignment has. I recommend `bowtie1` with `-a`, `--best`, and `--strata` parameters because it produces all best alignments that have the minimum number of mismatches per read. But I also wanted to design `emase` agnostic of alignment strategies for users like you, and  made `alntools` process any `bam` files. Why don't you reinstall `bowtie1`?

**Comment by mfrancesconi81** [comment]
Hi KB,

Sorry for not being clear. How should I install bowtie if the last version
in bioconda (1.2.2) gives me errors? Which version of bowtie are you using?

M.

**Comment by kbchoi-jax** [comment]
I tested up to v1.0.0. But I don't imagine the newer versions would fail. As I mentioned, `alntools` is aligner-agnostic.

**Comment by mfrancesconi81** [comment]
Hi KB,

I solved the bowtie problem by installing it from sourceforge instead of
bioconda.  But I still have trouble with the command alntools bam2ec.

I run the alntools bam2ec using emase.pooled.targets.info produced by
create-hybrid as transcript length file and the bam file produced by bowtie
using -a --best --strata -v -m 100 parameters as per your recommendation,
and -c 14 because I previously split the bam into 14 chuncks using alntools
split.

The program start running and immediately produces temporary files
(_bam2ec.0.bam, etc..) and output  "Unexpected target found in BAM file:" and
the name of the first transcript found in each of the split bam files. Then
it keeps running for more than 6 hours without producing any other output
or updating any of the temporary files.

I tried to terminate using ctrl c without success. I had to close the
session on the cluster.

Below is the head of the bam file.  A6 and A7 are the two haplotypes.  Does
it look good to you? Any clue on what might be going wrong?

Thank you very much,

Mirko

D00733:255:CB6DCANXX:7:1110:1899:2181 16 FBtr0302317_A7 1819 255 50M * 0 0
TACCAAGCCGAGCACTCTGGAAGAGCTGAAGCATCAGCCGGTTATAGTGG
GGGGGGGGGGGFGGFCEGGGGGGGGGGGGGGGGGGGGGGGGCGGGCCCCC XA:i:0 MD:Z:50 NM:i:0
XM:i:3
D00733:255:CB6DCANXX:7:1110:1899:2181 16 FBtr0083452_A7 1819 255 50M * 0 0
TACCAAGCCGAGCACTCTGGAAGAGCTGAAGCATCAGCCGGTTATAGTGG
GGGGGGGGGGGFGGFCEGGGGGGGGGGGGGGGGGGGGGGGGCGGGCCCCC XA:i:0 MD:Z:50 NM:i:0
XM:i:3
D00733:255:CB6DCANXX:7:1110:1992:2183 16 FBtr0086537_A6 497 255 50M * 0 0
AGTCGGGTGCCGGCAACAACTGGGCCAAGGGTCATTACACAGAGGGTGCC
EGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCBBB XA:i:1 MD:Z:49T0 NM:i:1
XM:i:5
D00733:255:CB6DCANXX:7:1110:1992:2183 16 FBtr0086536_A7 392 255 50M * 0 0
AGTCGGGTGCCGGCAACAACTGGGCCAAGGGTCATTACACAGAGGGTGCC
EGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCBBB XA:i:1 MD:Z:49T0 NM:i:1
XM:i:5
D00733:255:CB6DCANXX:7:1110:1992:2183 16 FBtr0086537_A7 497 255 50M * 0 0
AGTCGGGTGCCGGCAACAACTGGGCCAAGGGTCATTACACAGAGGGTGCC
EGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCBBB XA:i:1 MD:Z:49T0 NM:i:1
XM:i:5
D00733:255:CB6DCANXX:7:1110:1992:2183 16 FBtr0086536_A6 392 255 50M * 0 0
AGTCGGGTGCCGGCAACAACTGGGCCAAGGGTCATTACACAGAGGGTGCC
EGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCBBB XA:i:1 MD:Z:49T0 NM:i:1
XM:i:5
D00733:255:CB6DCANXX:7:1110:1816:2200 16 FBtr0110818_A7 1089 255 50M * 0 0
TGGAAAACAATGGAAATGCAACAGGAAAACGAGACGGGAAGAGAAGGGGC
GGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGBB<BA XA:i:1 MD:Z:49T0 NM:i:1
XM:i:9
D00733:255:CB6DCANXX:7:1110:1816:2200 16 FBtr0110818_A6 1089 255 50M * 0 0
TGGAAAACAATGGAAATGCAACAGGAAAACGAGACGGGAAGAGAAGGGGC
GGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGBB<BA XA:i:1 MD:Z:49T0 NM:i:1
XM:i:9
D00733:255:CB6DCANXX:7:1110:1816:2200 16 FBtr0333850_A6 995 255 50M * 0 0
TGGAAAACAATGGAAATGCAACAGGAAAACGAGACGGGAAGAGAAGGGGC
GGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGBB<BA XA:i:1 MD:Z:49T0 NM:i:1
XM:i:9
D00733:255:CB6DCANXX:7:1110:1816:2200 16 FBtr0333850_A7 995 255 50M * 0 0
TGGAAAACAATGGAAATGCAACAGGAAAACGAGACGGGAAGAGAAGGGGC
GGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGBB<BA XA:i:1 MD:Z:49T0 NM:i:1
XM:i:9
